---
title: "SWTH WGSassign"
author: "Christen Bossu"
date: "5/29/2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(tidyverse)
```

## Breeding and nonbreeding beagle file was thinned to 10 equally sized files by line using awk and labeled with aa-aj

Breeding beagle was subset into a reference bird beagle so each genetic group had similar effective sample size and non-reference beagle. Non-reference individuals would be assigned to breeding genetic groups similar to the nonbreeding birds

The breeding and nonbreeding beagle files can be found on dryad.

```{bash}
conda activate bcftools

zcat SWTH.breed.combined.subsetted.beagle.gz |awk 'NR % 10 == 0 { print $0 }' |cut -f 1-3,4-6,7-9,10-12,13-15,16-18,19-21,22-24,25-27,31-33,49-51,61-63,64-66,82-84,103-105,118-120,121-123,142-144,175-177,208-210,214-216,235-237,238-240,241-243,244-246,268-270,280-282,289-291,310-312,316-318,322-324,325-327,328-330,337-339,346-348,349-351 > REF35.breed.combined.subsetted.beagle.aa


zcat SWTH.breed.combined.subsetted.beagle.gz|awk 'NR % 10 == 0 { print $0 }' |cut -f 1-3,28-30,34-36,37-39,40-42,43-45,46-48,52-54,55-57,58-60,67-69,70-72,73-75,76-78,79-81,85-87,88-90,91-93,94-96,97-99,100-102,106-108,109-111,112-114,115-117,124-126,127-129,130-132,133-135,136-138,139-141,145-147,148-150,151-153,154-156,157-159,160-162,163-165,166-168,169-171,172-174,178-180,181-183,184-186,187-189,190-192,193-195,196-198,199-201,202-204,205-207,211-213,217-219,220-222,223-225,226-228,229-231,232-234,247-249,250-252,253-255,256-258,259-261,262-264,265-267,271-273,274-276,277-279,283-285,286-288,292-294,295-297,298-300,301-303,304-306,307-309,313-315,319-321,331-333,334-336,340-342,343-345,352-354,355-357 > nonref.breed.combined.subsetted.beagle.aa


zcat SWTH.nonbreed.combined.subsetted.beagle.gz| awk 'NR % 10 == 0 { print $0 }' > SWTH.nonbreed.combined.subsetted.beagle.thin_aa
```

The beagle was split by every line with awk:
```
awk 'NR % 10 == 0 { print $0 }'
awk 'NR % 10 == 1 { print $0 }'
awk 'NR % 10 == 2 { print $0 }'
awk 'NR % 10 == 3 { print $0 }'
awk 'NR % 10 == 4 { print $0 }'
awk 'NR % 10 == 5 { print $0 }'
awk 'NR % 10 == 6 { print $0 }'
awk 'NR % 10 == 7 { print $0 }'
awk 'NR % 10 == 8 { print $0 }'
awk 'NR % 10 == 9 { print $0 }'
```

Notes- one of these will have the header. Create a header file (yes a file with 1 line)

You can bgzip the file with the header immediately, otherwise create a for loop to concat the header file with the beagle file and compress it with bgzip

```{bash}
conda activate bcftools
for j in file.aa file.ac file.ad file.ae file.af file.ag file.ah file.ai file.aj
do
cat head "$j" |bgzip -c > "$j".gz
done
```

## Run WGSassign to get population allele frequency for each beagle block and test the effect sample size for the first block (aa). 

Cross-validation is an important technique for determining assignment accuracy. When producing reference population allele frequencies, you can add --loo to specify the calculation of leave-one-out assignment likelihoods for each individual to each reference population.

```{bash}
conda activate WGSassign

## Check the effective sample size (ESS)
WGSassign -b REF35.breed.combined.subsetted.beagle.aa.gz --pop_af_IDs SWTH.breeding.ind35.reference_K3.IDs.txt --get_reference_af  --out REF35.angsd.thin_10_aa.beagle-gl.ne --ne_obs

## Run leave one out cross validation (LOO) for each subset
for j in aa ab ac ad ae af ag ah ai aj
do
WGSassign -b REF35.breed.combined.subsetted.beagle."$j".gz --pop_af_IDs SWTH.breeding.ind35.reference_K3.IDs.txt --get_reference_af  --out REF35.angsd.thin_10_"$j".beagle-gl.LOO --loo
done


```

## Run WGSassign to produce the population log-likelihoods of assignment for each individual in each subsetted beagle file (aa-aj) to a reference population allele frequency file assignment 

```{bash}
for j in aa ab ac ad ae af ag ah ai aj
do
WGSassign --beagle SWTH.nonbreed.combined.subsetted.beagle.thin_"$j".gz --pop_af_file REF35.angsd.thin_10_"$j".beagle-gl.LOO.pop_af.npy --get_pop_like --out swth.133.subsetted.beagle.thin_"$j" --threads 1
done
```

Will also run the above for the nonreference breeding beagle file:  nonref.breed.combined.subsetted.beagle."$j".gz 

But the below code chunk requires specific files that have the order of individuals, so each have to be run separately, so WGSassign is run separately, then the the summary/conversion script.

## Convert WGSoutput to something readible in R

```{r}
## Read in files
files <- list.files(
  path = "output/WGSassign/SWTH",
  pattern = "^swth.133.*\\.pop_like.txt$", 
  full.names = TRUE
)
files

## Function to summarise WGSassign output. This needs to be done for reference breeders, nonreference breeders and nonbreeding birds separately by changing the ne_ind_order_sub file

## For reference breeders
ne_ind_order_sub<-read_delim("input/WGSassign/SWTH/REF35.ind.colnum.txt",delim="\t",col_names = T) %>% mutate(Dataset="LOO")

## For nonreference breeders
ne_ind_order_sub<-read_delim("input/WGSassign/SWTH/Nonref_breed.ind.colnum.txt",delim="\t",col_names = T) %>% mutate(Dataset="Individual")

## Create function
format_wgsout <- function(f) {
  breeding.likelihood <- read_delim(f, delim=" ",col_names = F)
  names(breeding.likelihood)<-c("BR", "MP", "NT", "ST","WB")

  # Process output
  ne_ind_order_sub<-read_delim("input/WGSassign/SWTH/AMRE_wintering_sample_table.tsv",delim="\t",col_names = T) %>% rename(Sample=sample_name) %>% dplyr::select(Sample)

loo.summary <- cbind(breeding.likelihood,ne_ind_order_sub) %>%
  pivot_longer(cols = BR:WB,
               names_to = "AssignedPop",
               values_to = "AssignedLike") %>%
  group_by(Sample) %>%
  mutate(AssignedProb = round(exp(AssignedLike - max(AssignedLike)) / sum(exp(AssignedLike - max(AssignedLike))),2 )) %>%
  filter(AssignedLike == max(AssignedLike)) %>%
  ungroup() 
write.table(loo.summary, paste(f, ".WGS_summary.txt", sep=""), row.names=F, col.names=T,sep="\t", quote=F)
}

lapply(files, format_wgsout)
```

#Thinned- not consecutive chunks of reference genome
```{r}
ne_ind_order_sub<-read_delim("input/WGSassign/SWTH/REF35.ind.colnum.txt",delim="\t",col_names = T) 

brloo.summary1 <- read_delim("input/WGSassign/SWTH/REF35.breed.combined.subsetted.beagle.aa.LOO.pop_like_LOO.txt.WGS_summary.txt",delim="\t") %>%
  dplyr::select(Sample,repunit,AssignedPop)

brloo.summary2 <- read_delim("input/WGSassign/SWTH/REF35.breed.combined.subsetted.beagle.ab.LOO.pop_like_LOO.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,repunit,AssignedPop)

brloo.summary3 <- read_delim("input/WGSassign/SWTH/REF35.breed.combined.subsetted.beagle.ac.LOO.pop_like_LOO.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,repunit,AssignedPop)

brloo.summary4 <- read_delim("input/WGSassign/SWTH/REF35.breed.combined.subsetted.beagle.ad.LOO.pop_like_LOO.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,repunit,AssignedPop)

brloo.summary5 <- read_delim("input/WGSassign/SWTH/REF35.breed.combined.subsetted.beagle.ae.LOO.pop_like_LOO.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,repunit,AssignedPop)

brloo.summary6 <- read_delim("input/WGSassign/SWTH/REF35.breed.combined.subsetted.beagle.af.LOO.pop_like_LOO.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,repunit,AssignedPop)

brloo.summary7 <- read_delim("input/WGSassign/SWTH/REF35.breed.combined.subsetted.beagle.ag.LOO.pop_like_LOO.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,repunit,AssignedPop)

brloo.summary8 <- read_delim("input/WGSassign/SWTH/REF35.breed.combined.subsetted.beagle.ah.LOO.pop_like_LOO.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,repunit,AssignedPop)

brloo.summary9 <- read_delim("input/WGSassign/SWTH/REF35.breed.combined.subsetted.beagle.ai.LOO.pop_like_LOO.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,repunit,AssignedPop)

brloo.summary10 <- read_delim("input/WGSassign/SWTH/REF35.breed.combined.subsetted.beagle.aj.LOO.pop_like_LOO.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,repunit,AssignedPop)

rbind(brloo.summary1,brloo.summary2,brloo.summary3,brloo.summary4,brloo.summary5,brloo.summary6,brloo.summary7,brloo.summary8,brloo.summary9,brloo.summary10) %>% group_by(Sample,repunit,AssignedPop) %>% summarise(Consistency=n()/10) %>% arrange(Sample,-Consistency) %>% 
  #filter(Consistency>0.79) %>% 
  left_join(ne_ind_order_sub) %>% 
  pivot_wider(names_from = AssignedPop, values_from = c(Consistency)) %>% replace(is.na(.), 0) %>% write.table("output/WGSassign/SWTH/SWTH.REF35.thin_10.consistent.wide.angsd_loco-pipe.txt",row.names=F,quote=F,sep="\t")
```

#thin every 10th snp starting from 1,2..10 for winter assignment
```{r}
assign.summary1 <- read_delim("input/WGSassign/SWTH/swth.133.subsetted.beagle.thin_aa.pop_like.txt.WGS_summary.txt",delim="\t") %>%
  dplyr::select(Sample,AssignedPop)

assign.summary2 <- read_delim("input/WGSassign/SWTH/swth.133.subsetted.beagle.thin_ab.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

assign.summary3 <- read_delim("input/WGSassign/SWTH/swth.133.subsetted.beagle.thin_ac.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

assign.summary4 <- read_delim("input/WGSassign/SWTH/swth.133.subsetted.beagle.thin_ad.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

assign.summary5 <- read_delim("input/WGSassign/SWTH/swth.133.subsetted.beagle.thin_ae.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

assign.summary6 <- read_delim("input/WGSassign/SWTH/swth.133.subsetted.beagle.thin_af.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

assign.summary7 <- read_delim("input/WGSassign/SWTH/swth.133.subsetted.beagle.thin_ag.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

assign.summary8 <- read_delim("input/WGSassign/SWTH/swth.133.subsetted.beagle.thin_ah.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

assign.summary9 <- read_delim("input/WGSassign/SWTH/swth.133.subsetted.beagle.thin_ai.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

assign.summary10 <- read_delim("input/WGSassign/SWTH/swth.133.subsetted.beagle.thin_aj.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

rbind(assign.summary1,assign.summary2,assign.summary3,assign.summary4,assign.summary5,assign.summary6,assign.summary7,assign.summary8,assign.summary9,assign.summary10) %>% group_by(Sample,AssignedPop) %>% summarise(Consistency=n()/10) %>% arrange(Sample,-Consistency) %>% 
  #filter(Consistency>0.79) %>% 
  pivot_wider(names_from = AssignedPop, values_from = c(Consistency)) %>% replace(is.na(.), 0) %>%  write.table("output/WGSassign/SWTH/swth.ind133.thin_10_all.assigned_pop.REF35.wide_consistent.angsd.loco-pipe.txt",row.names=F,quote=F,sep="\t")
```


# Nonreference breeding individual assignment with thinned beagle
```{r}
nrbr.summary1 <- read_delim("input/WGSassign/SWTH/nonref.breed.subsetted.beagle.thin_aa.pop_like.txt.WGS_summary.txt",delim="\t") %>%
  dplyr::select(Sample,AssignedPop)

nrbr.summary2 <- read_delim("input/WGSassign/SWTH/nonref.breed.subsetted.beagle.thin_ab.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

nrbr.summary3 <- read_delim("input/WGSassign/SWTH/nonref.breed.subsetted.beagle.thin_ac.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

nrbr.summary4 <- read_delim("input/WGSassign/SWTH/nonref.breed.subsetted.beagle.thin_ad.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

nrbr.summary5 <- read_delim("input/WGSassign/SWTH/nonref.breed.subsetted.beagle.thin_ae.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

nrbr.summary6 <- read_delim("input/WGSassign/SWTH/nonref.breed.subsetted.beagle.thin_af.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

nrbr.summary7 <- read_delim("input/WGSassign/SWTH/nonref.breed.subsetted.beagle.thin_ag.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

nrbr.summary8 <- read_delim("input/WGSassign/SWTH/nonref.breed.subsetted.beagle.thin_ah.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

nrbr.summary9 <- read_delim("input/WGSassign/SWTH/nonref.breed.subsetted.beagle.thin_ai.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)

nrbr.summary10 <- read_delim("input/WGSassign/SWTH/nonref.breed.subsetted.beagle.thin_aj.pop_like.txt.WGS_summary.txt",delim="\t") %>% dplyr::select(Sample,AssignedPop)


rbind(nrbr.summary1,nrbr.summary2,nrbr.summary3,nrbr.summary4,nrbr.summary5,nrbr.summary6,nrbr.summary7,nrbr.summary8,nrbr.summary9,nrbr.summary10) %>% group_by(Sample,AssignedPop) %>% summarise(Consistency=n()/10) %>% arrange(Sample,-Consistency) %>% 
  #left_join(meta) %>% 
   pivot_wider(names_from = AssignedPop, values_from = c(Consistency)) %>% replace(is.na(.), 0) %>% write.table("output/WGSassign/SWTH/SWTH.nonref.thin_10_all.assigned_pop.REF35.consistent.wide.loco-pipe.txt",row.names=F,quote=F,sep="\t")
```
